### PR TITLE
Report health detail

### PR DESCRIPTION
**Pull request by neiljerram**
This change surfaces more detail about the internal components, within a given process such as
Felix, that contribute to that process's overall liveness and readiness state.  If a process is
reporting non-live, or non-ready, and it isn't immediately obvious why, this extra detail should
indicate where the problem is.

Here's an example for Felix during startup, when it hasn't yet managed to connect to the datastore:
```
+---------------+---------+----------------+---------------------+--------+
|   COMPONENT   | TIMEOUT |    LIVENESS    |      READINESS      | DETAIL |
+---------------+---------+----------------+---------------------+--------+
| felix-startup | 0s      | reporting live | reporting non-ready |        |
| int_dataplane | 20s     | reporting live | reporting non-ready |        |
+---------------+---------+----------------+---------------------+--------+
```

On the internal health API, this is provided as the `Detail` field of the report returned by
`HealthAggregator.Summary()`.  It's also now possible for individual health reporters to fill in
`Detail` when they make a report, and that would be shown in the "DETAIL" column.  (But no existing
coding sets any per-reporter `Detail` yet, as this commit has only just created the `Detail` field.)

On the HTTP endpoints `/liveness` and `/readiness`, this detail is provided as the body of the HTTP response.

### CODE REVIEW

File: libcalico-go/lib/health/health.go [hunk 1]
```diff
@@ -144,27 +147,29 @@ func NewHealthAggregator() *HealthAggregator {
 		reporters:    map[string]*reporterState{},
 		httpServeMux: http.NewServeMux(),
 	}
-	aggregator.httpServeMux.HandleFunc("/readiness", func(rsp http.ResponseWriter, req *http.Request) {
-		log.Debug("GET /readiness")
+	genResponse := func(rsp http.ResponseWriter, quality string, state bool, detail string) {
```
Comment (caseydavenport) [inline libcalico-go/lib/health/health.go:150]: any reason we need this to be an inline function? Might be easier to read if we make it a proper function.

File: libcalico-go/lib/health/health.go [hunk 2]
```diff
@@ -177,19 +182,50 @@ func (aggregator *HealthAggregator) Summary() *HealthReport {
 	// In the absence of any reporters, default to indicating that we are both live and ready.
 	summary := &HealthReport{Live: true, Ready: true}
 
+	// Prepare a table to report detail.
+	var buf bytes.Buffer
+	table := tablewriter.NewWriter(&buf)
+	table.SetHeader([]string{"COMPONENT", "TIMEOUT", "LIVENESS", "READINESS", "DETAIL"})
```
Comment (caseydavenport) [inline libcalico-go/lib/health/health.go:188]: Does this table format decently in `kubectl describe pod` output? My assumption based on the rest of this is that writing to the response body will correctly show up in `kubectl describe`, which is awesome, but I'm wondering if it's human-readable in the describe output

**Review by caseydavenport — APPROVED** [review]
Looks good to me - a couple of questions but pretty minor.

I think the main thing I'm wondering is if this will show up nicely in `kubectl describe` or not.

**Comment by neiljerram** [comment]
@caseydavenport 

> I think the main thing I'm wondering is if this will show up nicely in `kubectl describe` or not.

Annoying, no!  See https://github.com/kubernetes/kubernetes/blob/master/pkg/probe/http/http.go#L138.  That area of code carefully reads the response body, but then **fails** to include in the failure case!

In the corresponding Exec case, the stdout+stderr **is** reported: https://github.com/kubernetes/kubernetes/blob/master/pkg/probe/exec/exec.go#L69.  So one approach would be for us to use an Exec wrapper instead of a direct HTTP probe.  We could also submit a PR to include the body in the HTTP failure case.  WDYT?

**Comment by neiljerram** [comment]
I just raised this one: https://github.com/kubernetes/kubernetes/pull/108918

File: libcalico-go/lib/health/health.go (hunk 1 above)
Comment (neiljerram) [reply libcalico-go/lib/health/health.go]: Good idea, done.

**Comment by caseydavenport** [comment]
> So one approach would be for us to use an Exec wrapper instead of a direct HTTP probe

I think we may do this anyway in order to remove the need for opening an additional port on the host (which might conflict with other host processes). 

> We could also submit a PR to include the body in the HTTP failure case. WDYT?

Seems like a nice thing to do!

**Comment by caseydavenport** [comment]
> We could also submit a PR to include the body in the HTTP failure case. WDYT?

Ah yeah, I see it was due to a security issue. I remember that one now..

**Review by caseydavenport — APPROVED** [review]

**Comment by neiljerram** [comment]
Thanks Casey, I'll merge this one now once CI is good.  I have taken more of a look at the `kubectl describe` question, but (a) it's complicated, and (b) this enhancement is still useful without the kubelet integration.  So I think that already justifies merging it, but here is more detail for the record...

1. The extra output here can only possibly go through to `kubectl describe` if liveness or readiness probes are configured in the manifest.  I probably will do that for Skimble, but a detailed health report could be useful even without that integration.

But assuming we do have that integration:

2. AFAICT, there's no way via `kubectl` of seeing output for **successful** probes.  So we're only talking about the failed probe case.
3. We can't get an HTTP probe to include output, because of security concerns as in the linked PR discussion.
4. But an Exec probe looks like it will include output, from code reading.  And I think the HTTP security concerns are valid, so we should generally config Exec probes instead.
5. I haven't yet put everything together to try to see a failed liveness event for Skimble, but I will keep working towards that, and report back here.

**Comment by neiljerram** [comment]
Just one CI issue: the Wireguard flake that we've had for ages.

**Pull request merged**